server/src/main/java/org/elasticsearch/repositories/IndexMetaDataGenerations.java

```diff
@@ -0,0 +1,207 @@
+        final Map<String, String> updatedIndexMetaHashes = new HashMap<>(hashes);
+        updatedIndexMetaHashes.putAll(newHashes);
+        updatedIndexMetaLookup.compute(snapshotId, (snId, lookup) -> {
+            if (lookup == null) {
+                return Map.copyOf(newLookup);
+            } else {
+                final Map<IndexId, String> updated = new HashMap<>(lookup);
+                updated.putAll(newLookup);
+                return Map.copyOf(updated);
+            }
+        });
+        return new IndexMetaDataGenerations(updatedIndexMetaLookup, updatedIndexMetaHashes);
+    }
+
+    /**
+     * Create a new instance with the given snapshot removed.
+     *
+     * @param snapshotId SnapshotId
+     * @return new instance without the given snapshot
+     */
+    public IndexMetaDataGenerations withRemovedSnapshot(SnapshotId snapshotId) {
+        final Map<SnapshotId, Map<IndexId, String>> updatedIndexMetaLookup = new HashMap<>(lookup);
+        updatedIndexMetaLookup.remove(snapshotId);
+        final Map<String, String> updatedIndexMetaDataHashes = new HashMap<>(hashes);
+        updatedIndexMetaDataHashes.keySet().removeIf(
+            k -> updatedIndexMetaLookup.values().stream().noneMatch(hashes -> hashes.containsValue(k)));
+        return new IndexMetaDataGenerations(updatedIndexMetaLookup, updatedIndexMetaDataHashes);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(hashes, lookup);
+    }
+
+    @Override
+    public boolean equals(Object that) {
+        if (this == that) {
+            return true;
+        }
+        if (that instanceof IndexMetaDataGenerations == false) {
+            return false;
+        }
+        final IndexMetaDataGenerations other = (IndexMetaDataGenerations) that;
+        return lookup.equals(other.lookup) && hashes.equals(other.hashes);
+    }
+
+    @Override
+    public String toString() {
+        return "IndexMetaDataGenerations{lookup:" + lookup + "}{hashes:" + hashes + "}";
+    }
+
+    /**
+     * Serialize and return the hex encoded SHA256 of {@link IndexMetaData}.
+     *
+     * @param indexMetaData IndexMetaData
+     * @return hex encoded SHA-256
+     */
+    public static String hashIndexMetaData(IndexMetaData indexMetaData) {
+        // Remove primary terms and version from the metadata to get a consistent hash across allocation changes
+        final IndexMetaData.Builder normalized = IndexMetaData.builder(indexMetaData).version(1L);
+        for (int i = 0; i < indexMetaData.getNumberOfShards(); i++) {
+            normalized.primaryTerm(i, 1L);
+        }
+        // IndexMetaData is serialized as a number of nested maps without any ordering guarantees. To get a consistent hash we first turn
+        // it into a nested tree-map to get ordering across all fields and then hash the nested tree-map
+        // TODO: Do the conversion of metadata to map more efficiently without the serialization round-trip
```

Comment:
This is a little horrible to tie the hashing to a serialization round-trip like this, but I didn't want to do the hash independently of the serialization functionality because of the maintenance nightmare that would cause